File: global/a_steadfast_servant.pl
#
# Steadfast Servant Veteran AA

$HPThreshold = 50;      # HP Threshold at which HoT will be cast.
$Frequency = 35;        # Frequency in seconds with which the EVENT_TIMER is executed.
$Range = 100;           # How close a player must be for a chance to be buffed.

@ManaRegenSpells = (697, 174, 1693, 6898);      # Breeze, Clarity, Clarity II, Servant's Breath
@HasteSpells = (10, 170, 171, 172);             # Augmentation, Alacrity, Celerity, Swift like the wind
@HoTSpells = (2502, 2175, 1283, 1522);          # Celestial Remedy,  Health, Cleansing, Elixir
@ACSpells = (18, 19, 20, 3470);                 # Guard, Armor of Faith, Shield of Words, Ward of Gallantry

sub EVENT_SPAWN {
        quest::settimer("RandomAction", $Frequency);
}

sub EVENT_TIMER {

        # First we find our owner and see if is in range and in need of any buffs.
        #
        $OwnerID = $npc->GetFollowID();

        if($OwnerID <= 0)
        {
                return;
        }
        $Owner = $entity_list->GetClientByID($OwnerID);

        if($Owner)
        {
                $x = $Owner->GetX();
                $y = $Owner->GetY();
                $z = $Owner->GetZ();

                $Distance = $npc->CalculateDistance($x, $y, $z);

                if($Distance < $Range) {
                        if(BuffPlayer($Owner))
                        {
                                # We cast a buff on our owner, so we are done for this cycle.
                                return;
                        }
                }
        }

        # Our owner didn't need any buffs, so look for someone else within range.
        #
        $c = $entity_list->GetRandomClient($x, $y, $z, $Range, $Owner);

        if(!$c)
        {
                return;
        }

        BuffPlayer($c);
}

sub BuffPlayer
{
        $c = $_[0];

        if($c)
        {
                $ClientID = $c->GetID();

                $HPRatio = ($c->GetHP() / $c->GetMaxHP()) * 100;

                $Level = $c->GetLevel();

                if($Level < 15)
                {
                        $SpellSet = 0;
                }
                elsif($Level < 30)
                {
                        $SpellSet = 1;
                }
                elsif($Level < 50)
                {
                        $SpellSet = 2;
                }
                else
                {
                        $SpellSet = 3;
                }

                $ManaRegenSpell = $ManaRegenSpells[$SpellSet];
                $HasteSpell = $HasteSpells[$SpellSet];
                $HoTSpell = $HoTSpells[$SpellSet];
                $ACSpell = $ACSpells[$SpellSet];

                if($HPRatio <= $HPThreshold)
                {
                        # Check if the player already has our HoT spell on.
                        #
                        $Slot = $c->FindBuff($HoTSpell);
                        if(!$Slot)
                        {
                                if($c->CanBuffStack($HoTSpell, $npc->GetLevel(), True) >= 0)
                                {
                                        $npc->FaceTarget($c);
                                        $npc->CastSpell($HoTSpell, $ClientID);
                                        return 1;
                                }
                        }
                }

                # Next the mana regen buff. Only cast this if the player uses mana.
                #
                if($c->GetMaxMana() > 0)
                {
                        $Slot = $c->FindBuff($ManaRegenSpell);
                        if(!$Slot)
                        {
                                if($c->CanBuffStack($ManaRegenSpell, $npc->GetLevel(), True) >= 0)
                                {
                                        $npc->FaceTarget($c);
                                        $npc->CastSpell($ManaRegenSpell, $ClientID);
                                        return 1;
                                }
                        }
                }

                # Finally, see if they need our Haste or AC buff.
                #
                @Buffs = ($HasteSpell, $ACSpell);

                foreach $Spell (@Buffs)
                {
                        $Slot = $c->FindBuff($Spell);
                        if(!$Slot)
                        {
                                if($c->CanBuffStack($Spell, $npc->GetLevel(), True) >= 0)
                                {
                                        $npc->FaceTarget($c);
                                        $npc->CastSpell($Spell, $ClientID);
                                        return 1;
                                }
                        }
                }
        }

        return 0;
} 

File: global/a_steadfast_servant.lua
local hp_threshold		= 50;						-- HP Threshold at which HoT will be cast.
local interval			= 35 * 1000;				-- Frequency in ms with which the timer is executed.
local max_distance		= 100;						-- How close a player must be for a chance to be buffed

local mana_regen_spells	= {697, 174, 1693, 6898};	-- Spells: Breeze, Clarity, Clarity II, Servant's Breath
local haste_spells		= {10, 170, 171, 172};		-- Spells: Augmentation, Alacrity, Celerity, Swift like the wind
local hot_spells		= {2502, 2175, 1283, 1522}; -- Spells: Celestial Remedy,  Health, Cleansing, Elixir
local ac_spells			= {18, 19, 20, 3470};		-- Spells: Guard, Armor of Faith, Shield of Words, Ward of Gallantry

function event_spawn(e)
	eq.set_timer("action", interval);
end

function event_timer(e)
	local pet_owner = e.self:GetOwner();

	if pet_owner.valid then
		local owner_x,owner_y,owner_z = pet_owner:GetX(), pet_owner:GetY(), pet_owner:GetZ();
		local owner_distance = e.self:CalculateDistance(owner_x,owner_y,owner_z) or 999999;

		if owner_distance < max_distance then
			if buff_player(e,pet_owner) then
				return;
			end
		end
	end

	-- Owner didn't need buffs, grab local client
	local random_player = eq.get_entity_list():GetRandomClient(e.self:GetX(),e.self:GetY(),e.self:GetZ(),max_distance);

	if random_player.valid then
		buff_player(e,random_player)
	end
end

function buff_player(e,client)
	if client.valid then
		local client_hp_ratio	= client:GetHPRatio();
		local client_level		= client:GetLevel();
		local spell_set			= 1;

		if client_level >= 50 then
			spell_set = 4;
		elseif client_level >= 30 then
			spell_set = 3
		elseif client_level >= 15 then
			spell_set = 2
		end

		if client_hp_ratio <= hp_threshold then
			if not client:FindBuff(hot_spells[spell_set]) then
				if client:CanBuffStack(hot_spells[spell_set],e.self:GetLevel(),true) then
					e.self:FaceTarget(client);
					e.self:CastSpell(hot_spells[spell_set], client:GetID());
					return true;
				end
			end
		end

		if client:GetMaxMana() > 0 then
			if not client:FindBuff(mana_regen_spells[spell_set]) then
				if client:CanBuffStack(mana_regen_spells[spell_set],e.self:GetLevel(),true) then
					e.self:FaceTarget(client);
					e.self:CastSpell(mana_regen_spells[spell_set], client:GetID());
					return true;
				end
			end
		end

		local buffs = {haste_spells[spell_set], ac_spells[spell_set]};

		for i = 1, #buffs do
			if not client:FindBuff(buffs[i]) then
				if client:CanBuffStack(buffs[i],e.self:GetLevel(),true) then
					e.self:FaceTarget(client);
					e.self:CastSpell(buffs[i], client:GetID());
					return true;
				end
			end
		end

		return false;
	end
end